### PR TITLE
configurable metrics (part 2): refactor dimensions to be a vector, permit overriding

### PR DESCRIPTION
**Pull request by kyessenov**
To be stacked upon #2640 
- [x] change dimensions to be a vector
- [x] add separate tags to each metric family and tag-to-vector index mapping
- [x] add overrides 
- [x] integration test
- [x] test for error cases
- [x] add metric definition

### CODE REVIEW

**Comment by kyessenov** [comment]
@gargnupur @mandarjog @bianpengyuan This is ready for review.
Two issues to be resolved:
- ABI is not returning errors, to be fixed by @jplevyak 
- creating new metrics dynamically not yet implemented

File: extensions/stats/plugin.cc [hunk 1]
```diff
@@ -55,45 +121,66 @@ void clearTcpMetrics(::Wasm::Common::RequestInfo& request_info) {
...
 
   // Seed the common metric tags with the default set.
-  tags_ = IstioDimensions::defaultTags();
-
-  // Process the dimension overrides
-  for (const auto& metric : config_.metrics()) {
-    if (metric.dimensions().empty()) {
-      continue;
+  std::vector<MetricFactory> factories = DefaultMetrics();
+  Map<std::string, std::vector<MetricTag>> metric_tags;
```
Comment (gargnupur) [inline extensions/stats/plugin.cc]: can string here be const ref?

File: extensions/stats/plugin.h [hunk 2]
```diff
@@ -316,33 +151,115 @@ class SimpleStat {
...
-  inline StringView name() const { return name_; };
-  inline bool is_tcp_metric() const { return is_tcp_metric_; }
+  inline StringView name() const { return metric_.name; };
+  inline bool is_tcp_metric() const { return is_tcp_; }
 
   // Resolve metric based on provided dimension values.
-  SimpleStat resolve(std::vector<std::string>& vals) {
-    auto metric_id = metric_.resolveWithFields(vals);
-    return SimpleStat(metric_id, value_fn_);
+  SimpleStat resolve(const IstioDimensions& instance) {
```
Comment (gargnupur) [inline extensions/stats/plugin.h]: Can we add comments to this function?

**Comment by mandarjog** [comment]
@kyessenov the other PR is not cherry-picked into 1.5 yet, are you going to just do 1 larger PR in 1.5 ?

File: extensions/stats/plugin.cc (hunk 1 above)
Comment (kyessenov) [reply extensions/stats/plugin.cc]: STD container don't like references, right? I think you cannot use a reference as key type. In any case, this is a temporary map as it gets destroyed on exit.

File: extensions/stats/plugin.h (hunk 2 above)
Comment (kyessenov) [reply extensions/stats/plugin.h]: Sure. The code is copy-pasted from ABI to avoid creating a temporary vector just to use the lower-level call.

**Comment by kyessenov** [comment]
Updated with the ABI change and added error test cases.
@mandarjog Yeah, I want to cherry pick both together in a PR.

**Comment by kyessenov** [comment]
Tests fail because of #2659

**Comment by kyessenov** [comment]
This is finishing the entire proposal (except for conditional matches, since that is kind of complicated and can be pushed into expressions if necessary).

File: extensions/stats/plugin.cc [hunk 3]
```diff
@@ -45,6 +45,76 @@ constexpr long long kDefaultTCPReportDurationNanoseconds = 15000000000;  // 15s
...
+}
+
+// maps peer_node and request to dimensions.
+void map(IstioDimensions& instance, bool outbound,
+         const wasm::common::NodeInfo& peer_node,
+         const ::Wasm::Common::RequestInfo& request) {
+  map_peer(instance, outbound, peer_node);
+  map_request(instance, request);
+  map_unknown_if_empty(instance);
+  if (request.request_protocol == "grpc") {
```
Comment (gargnupur) [inline extensions/stats/plugin.cc]: nit: can use kProtocolGRPC constant here..

File: extensions/stats/plugin.cc (hunk 1 above)
Comment (gargnupur) [reply extensions/stats/plugin.cc]: ahh okk it just allows const but no ref.. thanks!

**Pull request merged**